---
title: "Differential Expression Analysis"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library('TxDb.Mmulatta.UCSC.rheMac3.refGene')
library('biomaRt')
library('sleuth')
library('openxlsx')
library('ggplot2')
library('ggfortify')
library('matrixStats')
library('pheatmap')
library('WebGestaltR')
library('stringr')
library('orthogene')

source( "/ix/cigcore/utils/code/sleuth.func.R")
source("/ix/cigcore/utils/code/de.heatmap.R")

# Set paths
proj_path <- "/ix/cigcore/proj/mauricio"
data_path <- file.path(proj_path, "data")
result_path <- file.path(proj_path, "result")
experiment <- "03-18-2023-DE_Box1vs2vs3_covariate" # results with pvalue of 0.01
experiment_path <- file.path(result_path, experiment)
dir.create(experiment_path, recursive=TRUE)
```

```{r sleuth-setup, echo=FALSE, warning=FALSE, message=FALSE}
# Inputs for sleuth.func
metadataPath <- file.path(data_path, "sample_information.csv")
skipLines <- 0
sampleCol <- "Samples"

full_model <- "~Group + Subgroup"
factorCol <- c("Group", "Subgroup")

dataDirs <- c(file.path(data_path, "202212_OBS03"), file.path(data_path, "2022_Kiran/Mauricio Data"), 
                   file.path(data_path, "2021_fastq"))
outPath <- experiment_path

# mart <- useMart("ensembl", dataset="mmulatta_gene_ensembl") 
# ttg <- getBM(attributes=c("ensembl_gene_id", "ensembl_transcript_id", "external_gene_name"), mart = mart) 
# colnames(ttg) <- c("ensembl_gene", "target_id", "gene_name")
aggCol <- "ensembl_gene"
ttg <- read.csv(file.path(data_path, "ttg_rhesus.csv"))

pval <- 0.05

# pval <- 0.05
```

```{r run_sleuth.func_2, echo=FALSE, warning=FALSE, message=FALSE}
# Run sleuth.func with reference level==2
refLevel <- c("2", "E")
sleuth_results <- sleuth.func(metaData = metadataPath, skipLines=skipLines, sampleCol=sampleCol, full_model=full_model,
                                     factorCol = factorCol, refLevel=refLevel, dataDir = dataDirs,
                                     mappingDF = ttg, aggCol = aggCol, pvalCutoff = pval, outPath = outPath,
                                     drop_dupGenes = FALSE)

```

```{r sleuth-output-overall, echo=FALSE, message=FALSE, warning=FALSE}
# the following quantities are the same regardless of the reference level used in the sleuth analysis
norm_counts <- sleuth_results$normal_counts
sampleInfo <- sleuth_results$sampleDF

numSamples <- dim(sampleInfo)[1]
numTranscripts <- dim(norm_counts)[1]

tpmMat <- sleuth_results$TPM
write.csv(tpmMat, file.path(experiment_path, "tpm.csv"))
```

# Sample Metadata

```{r metadata, echo=FALSE}
knitr::kable(sampleInfo)
```

Adj. p-value: `r pval`

# Differential Expression Results

Number of samples: `r numSamples`

Total number of transcripts analyzed: `r numTranscripts`

## Reference Level = Group2

```{r Group2_summary, echo=FALSE}
summaryTable <- sleuth_results$summaryDF
knitr::kable(summaryTable)
```

### Group2 vs. Group1

#### PCA

```{r pca_2-1, echo=FALSE}
sleuth_results[["2_vs_1_pca"]]
```

```{r volcano_2-1, echo=FALSE, warning=FALSE, message=FALSE}
sleuth_results[["2_vs_1_vPlot"]]
```

### Group2 vs. Group3

#### PCA

```{r pca_2-3_pca, echo=FALSE}
sleuth_results[["2_vs_3_pca"]]
```

```{r volcano_2-3, echo=FALSE, warning=FALSE, message=FALSE}
sleuth_results[["2_vs_3_vPlot"]]
```

# Pathway Analysis

```{r webgestalt, include=FALSE, echo=FALSE, warning=FALSE, message=FALSE}
## Map rhesus ids to human
# markers_df <- read.xlsx("/ix/cigcore/proj/mauricio/result/DE_1_vs_2.xlsx")
# 
# ortho_df <- orthogene::convert_orthologs(markers_df, gene_input="ensembl_gene", 
#                                          gene_output = "columns", input_species="mmulatta",
#                                          output_species = "hsapiens", non121_strategy = "keep_both_species")

webgestalt_path <- file.path(experiment_path, "03-18-2023-pathway_analysis-covariate")
dir.create(webgestalt_path, recursive=TRUE)

resultsOut <- list(sleuth_results)
plots_list <- list()
mapDET <- list()

for(a1 in c("pathway_Reactome", "pathway_KEGG")){
  for (r1 in resultsOut){
    upOrDnGenes <- grep("\\w*(Up|Dn)\\b", names(r1), value=TRUE)
    for(u1 in upOrDnGenes){
      ## Find human orthologs
      ortho_df <- orthogene::convert_orthologs(r1[[u1]], gene_input="ensembl_gene", 
                                         gene_output = "columns", input_species="mmulatta",
                                         output_species = "hsapiens", non121_strategy = "keep_both_species")
      gene_list <- ortho_df$ortholog_gene
      projectName <- paste0(u1,"_",a1)
      mapDET[[u1]] <- ortho_df 
      wgr_out <- WebGestaltR(enrichMethod = "ORA", organism = "hsapiens", enrichDatabase = a1, 
                             interestGene = gene_list, interestGeneType = "genesymbol", referenceSet = "genome",
                             isOutput = TRUE, outputDirectory = webgestalt_path, projectName = projectName)
      
      comppath <- paste0("Project_", gsub("\\+", "_", projectName))     # account for fact that "+" not allowed in filenames
      docpath <- paste0("enrichment_results_", gsub("\\+", "_", projectName), ".txt")
      enrichpath <- file.path(webgestalt_path, comppath, docpath)
      
      if (file.exists(enrichpath)){
        enrichDF <- read.csv(enrichpath, sep ="\t", header = TRUE)
        enrichDF <- enrichDF[order(enrichDF$FDR), ]
        
        #rename rows/geneSets that have duplicated names
        unique_names <- make.names(enrichDF$description, unique = TRUE)
        unique_names <- gsub("\\.", "\\ ", unique_names)
        enrichDF$desc_unique <- unique_names
        
        enrichDF$desc_wrap <- str_wrap(enrichDF$desc_unique, width = 35)
        
        #barplot
        disp_rows <- min(40, dim(enrichDF)[[1]])
        plot1 <- ggplot(enrichDF[1:disp_rows,]) + geom_col(aes(x=enrichmentRatio, y=reorder(desc_wrap, enrichmentRatio))) + labs(y="treatment") + theme(axis.text = element_text(size = 5))
        plots_list[[projectName]] <- plot1
        ggsave(filename = file.path(webgestalt_path, comppath, "barplot.png"))
      }
    }
  }
}

write.xlsx(mapDET, file.path(experiment_path, "mapped_sigDET.xlsx"))
```

### Reference Level: Group2
#### Group2 vs. Group1 - Up-regulated (KEGG)
```{r enrichment_plot_2_vs_1-KEGG-Up, echo=FALSE, warning=FALSE, message=FALSE}
plots_list$sigDET_2_vs_1_Up_pathway_KEGG
```

#### Group2 vs. Group1- Down-regulated (KEGG)
```{r enrichment_plot_2_vs_1-KEGG-Dn, echo=FALSE, warning=FALSE, message=FALSE}
plots_list$sigDET_2_vs_1_Dn_pathway_KEGG
```

#### Group2 vs. Group1 - Up-regulated (Reactome)
```{r enrichment_plot_2_vs_1-Reactome-Up, echo=FALSE, message=FALSE, warning=FALSE}
plots_list$sigDET_2_vs_1_Up_pathway_Reactome
```

#### Group2 vs. Group1- Down-regulated (Reactome)
```{r enrichment_plot_2_vs_1-Reactome-Dn, echo=FALSE, warning=FALSE, message=FALSE}
plots_list$sigDET_2_vs_1_Dn_pathway_Reactome
```

#### Group2 vs. Group3 - Up-regulated (KEGG)
```{r enrichment_plot_2_vs_3-KEGG-Up, echo=FALSE, warning=FALSE, message=FALSE}
plots_list$sigDET_2_vs_3_Up_pathway_KEGG
```

#### Group2 vs. Group3- Down-regulated (KEGG)
```{r enrichment_plot_2_vs_3-KEGG-Dn, echo=FALSE, warning=FALSE, message=FALSE}
plots_list$sigDET_2_vs_3_Dn_pathway_KEGG
```

#### Group2 vs. Group3 - Up-regulated (Reactome)
```{r enrichment_plot_2_vs_3-Reactome-Up, echo=FALSE, message=FALSE, warning=FALSE}
plots_list$sigDET_2_vs_3_Up_pathway_Reactome
```

#### Group2 vs. Group3- Down-regulated (Reactome)
```{r enrichment_plot_2_vs_3-Reactome-Dn, echo=FALSE, warning=FALSE, message=FALSE}
plots_list$sigDET_2_vs_3_Dn_pathway_Reactome
```
